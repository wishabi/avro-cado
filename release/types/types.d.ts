/// <reference types="node" />
import * as Avro from "avsc";
export interface Options {
    schemaRegistry: string;
    numRetries: number;
    wrapUnions: string;
    subject: string;
    schema: object | null;
}
export interface EncoderTopicInfo {
    schema: Avro.Type | null;
    schemaId: number;
}
export interface EncodeFunc {
    (message: object): Buffer;
}
export declare type ResolverMap = {
    [index: number]: Promise<Avro.Resolver>;
};
export interface DecoderTopicInfo {
    subject: string;
    schema: Avro.Type | null;
    resolversMap: ResolverMap;
    retrieveSchema: Function;
    createSchemaResolver: Function;
    getSchemaResolver: Function;
}
export interface DecodeFunc {
    (buffer: Buffer): Promise<Object>;
}
