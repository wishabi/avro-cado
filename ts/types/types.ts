import * as Avro from "avsc";

/*
 ********************************************************
 *                                          PACKAGE TYPES
 ********************************************************
 */

// export interface Message {
//   topic: string;
//   value: Buffer;
//   offset: number;
//   partition: number;
//   key: Buffer;
//   size: number;
//   timestamp: number;
// }

export interface Options {
  schemaRegistry: string;
  numRetries: number;
  wrapUnions: string;
  subject: string;
  schema: any | null;
}

/*
 ********************************************************
 *                                          ENCODER TYPES
 ********************************************************
 */

export interface EncoderTopicInfo {
  schema: Avro.Type | null;
  schemaId: number;
}

export interface EncodeFunc {
  (message: object): Buffer;
}

/*
 ********************************************************
 *                                          DECODER TYPES
 ********************************************************
 */

export type ResolverMap = { [index: number]: Promise<Avro.Resolver> };

export interface DecoderTopicInfo {
  subject: string;
  schema: Avro.Type | null;
  resolversMap: ResolverMap;
  retrieveSchema: Function;
  createSchemaResolver: Function;
  getSchemaResolver: Function;
}

export interface DecodeFunc {
  (buffer: Buffer): Promise<Object>;
}
