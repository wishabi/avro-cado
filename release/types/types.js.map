{"version":3,"file":"types.js","sourceRoot":"","sources":["../../ts/types/types.ts"],"names":[],"mappings":""}