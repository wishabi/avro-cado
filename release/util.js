"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleError = (err, logger, msg) => {
    if (err.statusCode && err.statusCode === 500) {
        if (err.error &&
            err.error.error_code &&
            (err.error.error_code === 50003 || err.error.error_code === 50002)) {
            // retry
            logger.info(`retry ${msg}`, {
                statusCode: err.statusCode,
                errorCode: err.error.error_code,
                message: err.error.message
            });
            return true;
        }
    }
    logger.error(msg, err);
    return false;
};
exports.aggregateOptions = (defaultConf, overrideConf) => {
    const aggOptions = Object.assign({}, defaultConf, overrideConf);
    if (!overrideConf) {
        return aggOptions;
    }
    // determine how to handle unions
    const validWrapOptions = ["always", "never", "auto"];
    if (validWrapOptions.filter(opt => opt === aggOptions.wrapUnions).length === 0) {
        aggOptions.wrapUnions = defaultConf.wrapUnions;
    }
    return aggOptions;
};
//# sourceMappingURL=util.js.map