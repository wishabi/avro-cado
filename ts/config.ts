import { Options } from "../ts/types/types";
import { Logger } from "@flipp/flipp-node-logger";

export const ACCEPT_HEADERS: string =
  "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json";

export const optionsDefault: Options = {
  schemaRegistry: "http://localhost:8081",
  numRetries: 10,
  wrapUnions: "auto",
  subject: "DEFAULT",
  schema: null
};

export const encodeLogger: Logger = new Logger("encodeLogger");
export const decodeLogger: Logger = new Logger("decodeLogger");
