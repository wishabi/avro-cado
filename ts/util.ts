import { Options } from "./types/types";

export const RETRY_STATUS_CODE_500 = 500;
export const RETRY_ERROR_CODE_50002 = 50002;
export const RETRY_ERROR_CODE_50003 = 50003;

/**
 * Decide whether a schema registry error is eligible to
 * be retried
 *
 * @param err - the error in question
 * @return - true iff the error is eligible for retry
 */
export const handleError = (err: any): boolean => {
  if (
    err.statusCode &&
    err.statusCode === RETRY_STATUS_CODE_500 &&
    err.error &&
    err.error.error_code &&
    (err.error.error_code === RETRY_ERROR_CODE_50003 ||
      err.error.error_code === RETRY_ERROR_CODE_50002)
  ) {
    return true;
  }

  return false;
};

/**
 * Merge the user specified configuration options with the
 * default ones
 *
 * @param defaultConf - the default configuration options
 * @param overrideConf - the user specified configuration options
 * @return - merged configuration options object
 */
export const aggregateOptions = (
  defaultConf: Options,
  overrideConf: Options
): Options => {
  const aggOptions = Object.assign({}, defaultConf, overrideConf);

  if (!overrideConf) {
    return aggOptions;
  }

  // determine how to handle unions
  const validWrapOptions: Array<string> = ["always", "never", "auto"];
  if (
    validWrapOptions.filter(opt => opt === aggOptions.wrapUnions).length === 0
  ) {
    aggOptions.wrapUnions = defaultConf.wrapUnions;
  }

  return aggOptions;
};
