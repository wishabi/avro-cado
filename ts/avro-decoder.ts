import * as Avro from "avsc";
import * as rp from "request-promise";
import { ACCEPT_HEADERS, processOptions } from "./config";
import {
  DecodeFunc,
  DecoderInfo,
  Options,
  SchemaResolverFunc
} from "./types/types";
import { aggregateOptions, handleError } from "./util";

/**
 * Retrieve the Avro schema from the schema registry. On an error
 * eligible for retry, try up to the configure number of times
 *
 * @param subject - the subject under which the schema is registered
 * @param schemaRegistry - the schema registry path
 * @param numRetries - the number of retries
 * @param id - the schema id that should be retrieved
 * @return - The Avro schema requested
 */
export const retrieveSchema = async (
  { subject, schemaRegistry, numRetries }: Options,
  id: number
) => {
  const req = {
    uri: `${schemaRegistry}/schemas/ids/${id}`,
    headers: { Accept: ACCEPT_HEADERS },
    json: true,
    resolveWithFullResponse: true
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return JSON.parse((await rp(req)).body.schema);
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to retrieve schema for subject ${subject} with id ${id} :: ${
      error.message
    }`
  );
};

/**
 * Retrieve the Avro schema from the schema registry, parse it,
 * and create a resolver to the local Avro schema.
 *
 * @param id - the schema id that should be the source for the
 *             schema resolver
 * @param schema - the target Avro schema for the resolver
 * @return - The Avro schema resolver used to convert messages
 *           encoded with the specified Avro schema to the local
 *           Avro schema format
 */
export const genCreateSchemaResolver = (
  opts: Options
): SchemaResolverFunc => async (
  id: number,
  schema: Avro.Type
): Promise<Avro.Resolver> => {
  /**
   * Retrieve the schema by ID from the schema registry ...
   */
  const msgSchema = await retrieveSchema(opts, id);

  /**
   * ... parse the schema ...
   */
  try {
    const avSourceSchema = Avro.Type.forSchema(msgSchema);

    /**
     * ... and create a resolver to the schema we know how to consume
     */
    return schema.createResolver(avSourceSchema);
  } catch (err) {
    throw new Error(
      `Local schema is not compatible with schema from registry with id ${id} :: ${
        err.message
      }`
    );
  }
};

/**
 * Create an Avro decoder based on the specified parameters and
 * return a closure that takes an encoded Buffer and decodes it
 *
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param createSchemaResolver - a function used to create the resolver
 *                               object that will be used to decode the payload
 * @param resolversMap - a hashmap keys on the schema ids where the value is a
 *                       promise holding a schema resolver
 * @param buffer - The Avro encoded buffer
 * @return - The Avro decoded object that conforms to the specified schema
 */
export const genPayloadDecoder = ({
  schema,
  createSchemaResolver,
  resolversMap
}: DecoderInfo): DecodeFunc => async (
  buffer: Buffer | null
): Promise<object> => {
  // Do not attempt to avro decode null value
  if (!buffer) {
    return buffer;
  }

  /**
   * Extract the schema id with which the buffer was encoded
   */
  const id: number = buffer.slice(1, 5).readInt32BE(0);

  if (!resolversMap[id]) {
    /**
     * We do not have a resolver from the source schema to the
     * schema we support. Go and get a resolver.
     *
     * IMPORTANT
     *
     * We store a promise of a resolver in our hashmap. The reason is that
     * if we have multiple messages being handled asynchronously, we do not
     * want all of them to flood the schema registry with REST calls
     * to retrieve the schema
     */
    resolversMap[id] = createSchemaResolver(id, schema);
  }

  return schema.decode(buffer.slice(5), 0, await resolversMap[id]).value;
};

/*
 *****************************************************************
 *                      EXPORTED INTERFACE                       *
 *****************************************************************
 */

export const createDecoder = (opts: Options): DecodeFunc => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  // decoder info
  const decoderInfo: DecoderInfo = {
    subject: mergedOpts.subject,
    schema: Avro.Type.forSchema(mergedOpts.schema),
    resolversMap: {},
    createSchemaResolver: null
  };

  decoderInfo.createSchemaResolver = genCreateSchemaResolver(mergedOpts);

  return genPayloadDecoder(decoderInfo);
};
