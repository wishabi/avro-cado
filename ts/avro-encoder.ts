import * as Avro from "avsc";
import axios, { AxiosResponse } from "axios";
import { ACCEPT_HEADERS, processOptions } from "./config";
import { EncodeFunc, Options } from "./types/types";
import { handleError } from "./util";

/**
 * Register the specified schema for the specified topic under the
 * specified subject.
 *
 * On an error eligible for retry, keep trying up to the specified number
 * of times. Once all retries have been exhausted, any error is considered
 * a fatal one.
 *
 *
 * @param subject - the subject under which to register the schema
 * @param schemaRegistry - the schema registry host
 * @param numRetries - the number of retries to register before throwing an error
 * @param schema - the schema to register
 *
 * @return - A Promise holding the id of the schema in the schema registry
 */
export const registerSchema = async ({
  subject,
  schemaRegistry,
  schema,
  numRetries,
}: Options): Promise<number> => {
  // craft the REST call to the schema registry
  const req = {
    method: "POST",
    url: `${schemaRegistry}/subjects/${subject}/versions`,
    headers: { Accept: ACCEPT_HEADERS },
    data: {
      schema: JSON.stringify(schema),
    },
    json: true,
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return ((await axios(req)) as AxiosResponse).data.id;
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to register schema for subject ${subject} :: ${error.message}`,
  );
};

/**
 * Create a function that takes a message JSON object
 * and Avro encodes it
 *
 * @param payload - the JSON object to Avro encode
 * @return - The Avro encoded object as a Buffer
 */
export const genMessageEncoder = (
  schema: Avro.Type,
  schemaId: number,
): EncodeFunc => (payload: object): Buffer => {
  if (payload === null) {
    // no payload so return it
    return null;
  }

  // create the avro header
  const header = Buffer.alloc(5);

  // magic byte
  header.writeInt8(0, 0);

  // schema id
  header.writeInt32BE(schemaId, 1);

  // Avro encode the payload
  const buffer: Buffer = schema.toBuffer(payload);

  return Buffer.concat([header, buffer]);
};

/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/

export const createEncoder = async (opts: Options): Promise<EncodeFunc> => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  return genMessageEncoder(
    Avro.Type.forSchema(mergedOpts.schema, {
      wrapUnions: mergedOpts.wrapUnions,
    }),
    await registerSchema(mergedOpts),
  );
};
