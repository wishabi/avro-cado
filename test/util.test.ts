import { handleError, aggregateOptions } from "../ts/util";
import { logger } from "./__fixtures__/logger";
import { Options } from "../ts/types/types";

/*
 *****************************************************************
 *                                                     handleError
 *****************************************************************
 */

const RETRY_ERRORS = [
  {
    statusCode: 500,
    error: {
      error_code: 50002,
      message: "Retry error 500:50002"
    },
    message: "retry1"
  },
  {
    statusCode: 500,
    error: {
      error_code: 50003,
      message: "Retry error 500:50003"
    },
    message: "retry1"
  }
];

const FATAL_ERRORS = [
  {
    statusCode: 400,
    message: "message1"
  },
  { message: "message2" },
  {
    statusCode: 500,
    message: "message3"
  },
  {
    statusCode: 500,
    error: {
      message: "Retry error 500:50003"
    },
    message: "message4"
  }
];

describe("handleError", () => {
  it("should return 'true' for an error eligible for retry", () => {
    RETRY_ERRORS.forEach(retry_err => {
      const retry: boolean = handleError(retry_err, logger, retry_err.message);
      expect(retry).toMatchSnapshot();
    });
  });

  it("should return 'false' for an error NOT eligible for retry", () => {
    FATAL_ERRORS.forEach(fatal_err => {
      const retry: boolean = handleError(fatal_err, logger, fatal_err.message);
      expect(retry).toMatchSnapshot();
    });
  });
});

/*
 *****************************************************************
 *                                                aggregateOptions
 *****************************************************************
 */

const CONFIG_DATA = [
  {
    default: {
      schemaRegistry: "schemaRegistry_default1",
      numRetries: 3,
      wrapUnions: "auto"
    },
    override: {}
  },
  {
    default: {
      schemaRegistry: "schemaRegistry_default2",
      numRetries: 3,
      wrapUnions: "auto"
    },
    override: {
      schemaRegistry: "schemaRegistry_override2",
      numRetries: 3,
      wrapUnions: "auto"
    }
  },
  {
    default: {
      schemaRegistry: "schemaRegistry_default3",
      numRetries: 3,
      wrapUnions: "auto"
    },
    override: {
      schemaRegistry: "schemaRegistry_override3",
      numRetries: 4,
      wrapUnions: "always"
    }
  },
  {
    default: {
      schemaRegistry: "schemaRegistry_default4",
      numRetries: 3,
      wrapUnions: "auto"
    },
    override: {
      schemaRegistry: "schemaRegistry_default4",
      numRetries: 3,
      wrapUnions: "auto",
      topic: "topic_override"
    }
  },
  {
    default: {
      schemaRegistry: "schemaRegistry_default5",
      numRetries: 3,
      wrapUnions: "auto"
    },
    override: null
  },
  {
    default: {
      schemaRegistry: "schemaRegistry_default6",
      numRetries: 3,
      wrapUnions: "auto"
    },
    override: {
      schemaRegistry: "schemaRegistry_default6",
      numRetries: 3,
      wrapUnions: "petre"
    }
  }
];

describe("aggregateOptions", () => {
  it("should merge the two objects correctly", () => {
    CONFIG_DATA.forEach(test => {
      const merged = aggregateOptions(
        test.default as Options,
        test.override as Options
      );
      expect(merged).toMatchSnapshot();
    });
  });
});
