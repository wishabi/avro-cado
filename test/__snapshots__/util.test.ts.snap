// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aggregateOptions should merge the two objects correctly 1`] = `
Object {
  "numRetries": 3,
  "schemaRegistry": "schemaRegistry_default1",
  "wrapUnions": "auto",
}
`;

exports[`aggregateOptions should merge the two objects correctly 2`] = `
Object {
  "numRetries": 3,
  "schemaRegistry": "schemaRegistry_override2",
  "wrapUnions": "auto",
}
`;

exports[`aggregateOptions should merge the two objects correctly 3`] = `
Object {
  "numRetries": 4,
  "schemaRegistry": "schemaRegistry_override3",
  "wrapUnions": "always",
}
`;

exports[`aggregateOptions should merge the two objects correctly 4`] = `
Object {
  "numRetries": 3,
  "schemaRegistry": "schemaRegistry_default4",
  "topic": "topic_override",
  "wrapUnions": "auto",
}
`;

exports[`aggregateOptions should merge the two objects correctly 5`] = `
Object {
  "numRetries": 3,
  "schemaRegistry": "schemaRegistry_default5",
  "wrapUnions": "auto",
}
`;

exports[`aggregateOptions should merge the two objects correctly 6`] = `
Object {
  "numRetries": 3,
  "schemaRegistry": "schemaRegistry_default6",
  "wrapUnions": "auto",
}
`;

exports[`handleError should return 'false' for an error NOT eligible for retry 1`] = `false`;

exports[`handleError should return 'false' for an error NOT eligible for retry 2`] = `false`;

exports[`handleError should return 'false' for an error NOT eligible for retry 3`] = `false`;

exports[`handleError should return 'false' for an error NOT eligible for retry 4`] = `false`;

exports[`handleError should return 'true' for an error eligible for retry 1`] = `true`;

exports[`handleError should return 'true' for an error eligible for retry 2`] = `true`;
